File: lecture-1/2-bug2.c
#include<stdlib.h>
#include<stdio.h>
#include<string.h>

void print(char * msg) {
    static char * last;

    if (msg == NULL) {
        printf("%s\n", last);
    } else {
        printf("%s\n", msg);
        last = msg;
    }
}

int main(int argc, char ** argv) {
    char * user = malloc(5 * sizeof(char));
    memcpy(user, "user", 5);

    print(user);

    free(user);

    char * pass = malloc(5 * sizeof(char));
    memcpy(pass, "pass", 5);

    print(NULL);
    print(pass);
}

File: lecture-2/2-bug2.rs
// static mut global_last: &'static str = "";

fn main() {
    // global_last = "hello";
    let mut last : &String;

    // block 1
    {
        let user = "user".to_string();
        // print(user)
        {
            let msg = &user;
            println!("{}", msg);
            last = msg;
        }
        // print(user)
        {
            let msg = &user;
            println!("{}", msg);
            last = msg;
        }
    }
    // block 2
    {
        let pass = "pass".to_string();
        // print(NULL)
        {
            println!("{}", last);
        }
        // print(pass)
        {
            let msg = &pass;
            println!("{}", msg);
            last = msg;
        }
    }
}
